feat(pg-delta)!: clean-room rewrite, promoted to @supabase/pg-delta (BREAKING alpha) (#299)
supabase/pg-delta: packages/pg-delta/API-REVIEW.md

| `FactKind` | type | Union of all `StableId["kind"]` strings. Useful for narrowing/mapping. | ✓ | — | — | — |
| `encodeId` | function | `StableId → string` — the canonical string codec. Only place string encoding exists (guardrail 1). | ✓ | — | — | — |
| `parseId` | function | `string → StableId` — inverse of `encodeId`. | ✓ | — | — | — |
| `Payload` | type | `{ [key: string]: PayloadValue }` — the identity-free content of a fact. | ✓ | — | — | — |
| `ContentHash` | type | Opaque brand for a SHA-256 hex string produced by `contentHash`. | ✓ | — | — | — |
| `canonicalize` | function | Canonical deterministic JSON serialization of a `PayloadValue` — the equality surface of the whole system. | ✓ | — | — | — |
| `contentHash` | function | `PayloadValue → ContentHash` — SHA-256 of the canonical form. | ✓ | — | — | — |
| `Fact` | type | `{ id: StableId; parent?: StableId; payload: Payload }` — the atomic unit of schema knowledge. | ✓ | ✓ | ✓ | ✓ |
| `DependencyEdge` | type | `{ from, to: StableId; kind: EdgeKind }` — a directed dependency between two facts (tear-down / build-up order). | ✓ | ✓ | ✓ | — |

> | `EdgeKind` | type | `"depends" \| "owner" \| "memberOfExtension"` — semantic type of a `DependencyEdge`. | ✓ | — | — | — |

Reviewer: ```suggestion
| `EdgeKind` | type | `"depends" \| "owner" \| "memberOfExtension" \| "managedBy"` — semantic type of a `DependencyEdge`. | ✓ | — | — | — |
```

| `serializeSnapshot` | function | `(FactBase, { pgVersion }) → string` — format-v1 bigint-safe JSON, includes digest. | ✓ | — | — | — |
| `deserializeSnapshot` | function | `string → { factBase: FactBase; pgVersion: string }` — verifies digest on load; throws on corruption. | ✓ | — | — | — |
| `Delta` | type | Tagged union: `add / remove / set / link / unlink` — one atomic change between two fact bases. The diff vocabulary. | — | ✓ | — | — |
| `diff` | function | `(a: FactBase, b: FactBase) → Delta[]` — rollup-guided, zero per-kind code, deterministically sorted. | — | ✓ | — | — |

---

## Extract

| Name | Kind | Contract | facts | deltas | actions | proof |
|------|------|----------|:-----:|:------:|:-------:|:-----:|
| `ExtractResult` | type | `{ factBase: FactBase; pgVersion: string; diagnostics: Diagnostic[] }` — everything `extract` returns. | ✓ | — | — | — |
| `extract` | function | `Pool → Promise<ExtractResult>` — single REPEATABLE READ snapshot of a live database into a `FactBase`. | ✓ | — | — | — |

---

## Plan

| Name | Kind | Contract | facts | deltas | actions | proof |
|------|------|----------|:-----:|:------:|:-------:|:-----:|

> | `ENGINE_VERSION` | const | `"0.1.0"` — stamped into plan artifacts; `apply` refuses artifacts from other engines. | — | — | ✓ | — |

Reviewer: ```suggestion
| `ENGINE_VERSION` | const | Current engine compatibility version stamped into plan artifacts; `apply` refuses artifacts from other engines. | — | — | ✓ | — |
```

| `Action` | type | One executable DDL statement with `sql`, `verb`, `produces/consumes/destroys/releases`, `transactionality`, `lockClass`, `dataLoss`, `rewriteRisk`. The unit the executor runs. | — | — | ✓ | ✓ |

> | `Plan` | type | `{ actions, deltas, filteredDeltas, renameCandidates, safetyReport, source/target fingerprints, … }` — the complete output of the planner. | ✓ | ✓ | ✓ | ✓ |

Reviewer: ```suggestion
| `Plan` | type | Complete planner artifact: actions, managed/filtered deltas, optional compatibility `projectionAudit?`, policy/profile/scope metadata, rename decisions, safety report, and source/target fingerprints. Newly produced plans carry the audit; optionality permits legacy v1 parsing. | ✓ | ✓ | ✓ | ✓ |
```

| `SafetyReport` | type | Aggregated per-plan counts: destructive, rewriteRisk, nonTransactional actions; lock class histogram. | — | — | ✓ | — |
| `PlanOptions` | type | `{ params?, policy?, baseline?, renames?, acceptRenames?, compact?, capability? }` — the complete option bag for `plan()`. `baseline?` is the resolved platform `FactBase` whose facts are subtracted before diffing; `capability?` is the `ApplierCapability` that projects out operations the applier cannot execute. | — | — | ✓ | — |

> | `plan` | function | `(source, desired: FactBase, options?: PlanOptions) → Plan` — the planner: deltas × rule table → topologically sorted actions. | ✓ | ✓ | ✓ | — |

Reviewer: ```suggestion
| `plan` | function | `(source, desired: FactBase, options?: PlanOptions) → Plan` — the planner: deltas × rule table → topologically sorted actions. | ✓ | ✓ | ✓ | — |
| `ProjectionAudit` | type | Complete attributed raw differences hidden by managed-view projection: ordered `entries` plus recomputed suspicious/acknowledged/baseline summary counts. | ✓ | ✓ | — | ✓ |
| `ProjectionAuditEntry` | type | One hidden `Delta`, its fact/edge subject, all source/desired suppression causes, and the derived classification. | ✓ | ✓ | — | ✓ |
| `ProjectionAuditSuppression` | type | Stable attribution for one suppression: side, stage, reason code, classification, and optional excluded ancestor. | ✓ | ✓ | — | ✓ |
| `ProjectionAuditStage` | type | Projection boundary that caused suppression: `baseline`, `policyScopeRule`, `capability`, `managementScope`, `referenceOnly`, or `managedBy`. | ✓ | ✓ | — | ✓ |
| `ProjectionAuditClassification` | type | `"acknowledged" \| "suspicious"`; entry classification is recomputed from its suppressions at artifact/API boundaries. | — | ✓ | — | ✓ |
| `ProjectionAuditSubject` | type | Stable fact or dependency-edge identity duplicated from the delta for grouping and allowlisting. | ✓ | ✓ | — | ✓ |
```

| `LockClass` | type | `"none" \| "share" \| "shareRowExclusive" \| "shareUpdateExclusive" \| "accessExclusive"` — documented lock level of a DDL statement. Reported, not certified. | — | — | ✓ | — |

---

## Apply

| Name | Kind | Contract | facts | deltas | actions | proof |
|------|------|----------|:-----:|:------:|:-------:|:-----:|
| `ActionStatus` | type | `"applied" \| "unapplied" \| "inDoubt"` — per-action outcome after execution. | — | — | ✓ | ✓ |
| `ApplyOptions` | type | `{ fingerprintGate?, lockTimeoutMs?, statementTimeoutMs? }` — executor configuration. | — | — | ✓ | — |
| `ApplyReport` | type | `{ status, appliedActions, actionStatuses, error? }` — execution outcome with per-action status and a structured error entry. | — | — | ✓ | ✓ |
| `apply` | function | `(Plan, Pool, options?) → Promise<ApplyReport>` — sequential, segmented, lock-aware execution. | — | — | ✓ | ✓ |

---

## Proof

| Name | Kind | Contract | facts | deltas | actions | proof |
|------|------|----------|:-----:|:------:|:-------:|:-----:|
| `ProofVerdict` | type | `{ ok, applyError?, driftDeltas, dataViolations }` — full proof result including state check and data preservation. | ✓ | ✓ | ✓ | ✓ |

> | `provePlan` | function | `(Plan, clonePool: Pool, desired: FactBase) → Promise<ProofVerdict>` — apply to sacrificial clone, re-extract, diff against desired, check row counts. | ✓ | ✓ | ✓ | ✓ |

Reviewer: ```suggestion
| `ProofVerdict` | type | Public compatibility shape for consumer-authored verdicts. Includes proof failures/coverage plus optional `projectionAudit?`, `projectionAuditStatus?`, and `strictAuditFailure?` additive fields. | ✓ | ✓ | ✓ | ✓ |
| `ProducedProofVerdict` | type | Required result shape from `provePlan`: complete normalized `projectionAudit` and explicit `projectionAuditStatus: "available" \| "unavailable"`, plus all state/data/rewrite/coverage fields. | ✓ | ✓ | ✓ | ✓ |
| `ProveOptions` | type | Proof configuration: strict projection-audit enforcement, optional auto-seeding, managed-view policy/capability/baseline inputs, and a profile-aware re-extractor. | ✓ | ✓ | ✓ | ✓ |
| `provePlan` | function | `(Plan, clonePool: Pool, desired: FactBase, options?) → Promise<ProducedProofVerdict>` — applies to a sacrificial clone, proves state/data/rewrite claims, always returns audit status/data, and optionally makes suspicious or unavailable audits blocking through `strictAudit`. | ✓ | ✓ | ✓ | ✓ |
```
